Intake-ESM Integration based on #1218

charles-turner-1 opened the pull request:
## Description


***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [ ] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [ ] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [ ] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [ ] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [ ] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [ ] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

charles-turner-1 committed:
Add recognised intake-esm datastores on NCI systems to config_developer.yml, skeleton of intake-esm inclusiion following #1218

charles-turner-1 committed:
Skeleton

charles-turner-1 committed:
Playing around

charles-turner-1 committed:
Almost at a working IntakeDataset.load()

charles-turner-1 committed:
Working intake-esm implementation - probably still some kinks to iron out

charles-turner-1 committed:
Working with multiple catalogues per project

charles-turner-1 committed:
Cleanup - mypy & ruff errors

charles-turner-1 committed:
Remove WIP

charles-turner-1 committed:
Update depenencies & dev environment

charles-turner-1 committed:
Pre-commit modifications

charles-turner-1 committed:
Merge branch 'main' into intake-esm

charles-turner-1 committed:
Fixed most of codacy (mypy-strict?) gripes

charles-turner-1 committed:
Fix typo

bouweandela commented on esmvalcore/intake/_dataset.py:
```diff
@@ -0,0 +1,164 @@
+    return ([_CACHE[cat_url] for cat_url in catalog_urls], facet_list)
+
+
+class IntakeDataset(Dataset):
```
I'm having some reservations about subclassing the Dataset class for this purpose:
- A typical use case for many of our users will be that they have most data available from a central catalog that is managed by a central administrator, but want to augment that with the ability to download some files themselves. In that case, it is really useful to have the ability to deduplicate (e.g. pick the latest version of a file). I'm not sure if this can be achieved by subclassing the Dataset object.
- We will likely want to add support for other catalogs as well, e.g. intake-esgf, xcube, and STAC. If we need a new Dataset class for each of these, it may become confusing to users. 
- How will this work from the recipe?

As an alternative, would it be an option to register the available data sources in the configuration / `Dataset.session` and then make the `Dataset.files` method loop over the available sources and deduplicate?

bouweandela reviewed (commented):
Great to see progress on this @charles-turner-1!

bouweandela commented:
> Is there any documentation on where to find all potential facets that ESMValCore might accept & what they represent?

ESMValCore is quite flexible with what facets it accepts. We have a translation between some of 'our' facets and the official ones in the [`esmvalcore.esgf.facets`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.esgf.html#module-esmvalcore.esgf.facets) module. A few facets are used by ESMValCore for specific purposes such as CMOR checks and fixes (off the top of my head that would be `dataset`, `project`, `mip`, `short_name`), but others are entirely free-form and only used for finding input files and defining the output file names using the paths described in the config-developer.yml file.

> Our intake catalogs here on Gadi have a bunch of extra keys (facets) that I haven't mapped. 

If these are completely determined by the other facets, you can add them automatically using the [extra facets facility](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/quickstart/configure.html#extra-facets)

bouweandela commented:
> Structure: I've put this in an intake submodule,

How about adding a new module called e.g. `esmvalcore.data` or `esmvalcore.data_sources` or something similar and adding it as a submodule there? We could also move the `esmvalcore.local` and `esmvalcore.esgf` modules there (does not have to be in this pull request). I foresee us adding multiple input data sources in the near future.

charles-turner-1 commented:
Thanks for the review Bouwe, super helpful! I've only had a skim so far, but I'll get those suggestions incorporated next week

charles-turner-1 committed:
Beginning to work on Bouwe's comments (WIP)

charles-turner-1 committed:
Updates - restructured esmvalcore/data/intake following Bouwe's suggestions

charles-turner-1 committed:
Reorder imports (ruff maybe?)

charles-turner-1 committed:
Add `_read_facets` to intake configuration: see https://github.com/intake/intake-esm/pull/717

charles-turner-1 committed:
Add `merge_intake_seach_history` function (see https://github.com/intake/intake-esm/pull/717/commits/73f150e33724010a98b6aed15a39b2c09676a442)

charles-turner-1 committed:
Merge branch 'main' into intake-esm

bouweandela commented:
I started working on adding some interface code that could be useful here too in #2765.

valeriupredoi committed:
Merge branch 'main' into intake-esm

valeriupredoi commented:
@charles-turner-1 I popped the latest `main` here, that includes #2765 - do you reckon you'll have time to restart the work on it soon, mate? If not, no biggie, just pls let @bouweandela and myself know - we can take it from here, there is a bit of a tight schedule on getting full Zarr support (not only as a simple load via esmvalcore IO), and I reckon this is superuseful towards that :beers:

valeriupredoi committed:
readd intake

charles-turner-1 committed:
Merge branch 'main' into intake-esm

charles-turner-1 committed:
Add `data.io.intake_esm.py, scaffold off `data.io.intake_esgf.py`

charles-turner-1 committed:
WIP

charles-turner-1 committed:
Scaffold tests

charles-turner-1 committed:
Remove /data/intake stuff, /config/_intake

charles-turner-1 committed:
Pre-commit

charles-turner-1 committed:
Merge branch 'main' into intake-esm

charles-turner-1 committed:
Nearly there I think - all tests passing. Hopefully CI can tell me what I've missed :sweat_smile:

charles-turner-1 committed:
Pre-commit

charles-turner-1 committed:
Remove old intake-esm file

charles-turner-1 committed:
Merge branch 'main' into intake-esm

charles-turner-1 committed:
Merge branch 'main' into intake-esm

charles-turner-1 committed:
Sort keys when finding data - should guarantee order stability

charles-turner-1 committed:
Change path import style to match `/tests/integration/preprocessor/_io/_test_zarr` - might fix discovery?

charles-turner-1 committed:
Revert ugly type ignore stuff

charles-turner-1 committed:
Un-ignore the intake-esm data ncfiles

charles-turner-1 commented:
Few lines of coverage to fix, but I think this is mostly ready for review now!

charles-turner-1 committed:
Merge branch 'main' into intake-esm

charles-turner-1 committed:
- Pass through quiet arg correctly, remove special time range parsing - no capacity for this in intake-esm
- Add test for warning on empty search

bouweandela commented:
Great to see this progressing @charles-turner-1! Would it be possible to add some configuration files for using this, e.g. with the catalogs at your own institute, [Pangeo](https://intake-esm.readthedocs.io/en/stable/tutorials/loading-cmip6-data.html#) or [DKRZ cloud storage](https://tutorials.dkrz.de/tutorial_intake-1-introduction.html)?

charles-turner-1 committed:
Merge branch 'main' into intake-esm

charles-turner-1 committed:
Remove comment

charles-turner-1 committed:
Merge branch 'intake-esm' of github.com:ESMValGroup/ESMValCore into intake-esm

bouweandela commented:
I came across a CORDEX catalog at https://euro-cordex.s3.eu-central-1.amazonaws.com/catalog/CORDEX-CMIP5.json too: https://py-cordex.readthedocs.io/en/latest/notebooks/prudence.html#Investigate-regional-warming-in-Prudence-regions

charles-turner-1 committed:
Remove line subsetting variables - we're using paths anyway, so subsetting catalog is not helpful

charles-turner-1 committed:
Add a `time_separator` field (needed for flexibility in how datastoers are serialised

charles-turner-1 committed:
Allow catalog into constructor

charles-turner-1 committed:
Add sample config

charles-turner-1 committed:
Merge branch 'main' into intake-esm

charles-turner-1 committed:
Add description to intake-esm config file

charles-turner-1 committed:
Merge branch 'intake-esm' of github.com:ESMValGroup/ESMValCore into intake-esm

charles-turner-1 committed:
Merge branch 'main' into intake-esm

bouweandela commented:
The idea behind the configuration is that users use the command `esmvaltool config list` to explore which configurations are available and `esmvaltool config copy` to install the data source configurations that they want to use. How about using a prefix `data-intake-esm` as a prefix and then add some suffix that describes where the data is located?
- `data-intake-esm-aws` for data stored at Amazon (https://registry.opendata.aws/cmip6/, https://aws.amazon.com/marketplace/pp/prodview-unpqiiv5dfu6y, https://aws.amazon.com/marketplace/pp/prodview-mlahps6lrdmd4)
- `data-intake-esm-gcs` for data stored at Google (https://console.cloud.google.com/marketplace/product/noaa-public/cmip6?hl=en-GB)
- `data-intake-esm-dkrz-cloud` for data stored at DKRZ and accessible from outside DKRZ (https://tutorials.dkrz.de/tutorial_intake-1-introduction.html)
- `data-intake-esm-dkrz-disk` for data stored at DKRZ and accessible only from DKRZ machines (https://tutorials.dkrz.de/tutorial_intake-1-introduction.html)
- etc

It would be nice to perhaps add a subsection "Cloud" to https://docs.esmvaltool.org/projects/ESMValCore/en/latest/quickstart/configure.html#data-sources, similar to "Personal Computer" and explain how to set up data sources when you're working on a cloud computing system and from there link to slighly more extensive documentation in the module `esmvalcore.io.intake_esm`, similar to how it's done for https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.io.intake_esgf.html#module-esmvalcore.io.intake_esgf

charles-turner-1 committed:
Merge branch 'main' into intake-esm

charles-turner-1 committed:
- Update tests & functionality to use pangeo cloud catalog and pass through opening kwargs to intake-esm where necessary
- Add gcs to CI environmnet (not a hard dependency so only
  environment.yml, not pyproject.toml)
- Also remove a bunch of unnecessary test files

charles-turner-1 committed:
Update data-intake-esm config.yml & quickstart config docs

charles-turner-1 committed:
Add configs for various locations

charles-turner-1 committed:
update docs for intake-esm

charles-turner-1 committed:
Remove duplicate key

charles-turner-1 commented:
I've been running through a bunch of datasets in the Pangeo GCP intake-esm zarr datastore for tests, and variously running into a bunch of errors - I assume these are because of the tighter data model in `iris` as compared to `xarray`? These two are caught with `pytest.raises` in the unit tests, but I can't imagine we'd want to let such errors pass through...

eg. 
```python
ValueError: When encoding chunked arrays of datetime values, both the units and dtype must be prescribed or both must be unprescribed. Prescribing only one or the other is not currently supported. Got a units encoding of hours since 0151-01-16 12:00:00.000000 and a dtype encoding of None.
Raised while encoding variable 'time_bnds' with value <xarray.Variable (time: 6000, bnds: 2)> Size: 96kB
dask.array<where, shape=(6000, 2), dtype=object, chunksize=(6000, 2), chunktype=numpy.ndarray>
Attributes:
    long_name:  time axis boundaries
```
and 
```python
*** KeyError: 'O'
Traceback (most recent call last):
  File "/Users/u1166368/ESMValCore/esmvalcore/io/intake_esm.py", line 142, in to_iris
    return dataset_to_iris(dataset)
  File "/Users/u1166368/ESMValCore/esmvalcore/iris_helpers.py", line 628, in dataset_to_iris
    cubes = conversion_func(dataset)
  File "/Users/u1166368/miniforge3/envs/esmvaltool/lib/python3.14/site-packages/ncdata/iris_xarray.py", line 54, in cubes_from_xarray
    cubes = to_iris(ncdata, **iris_load_kwargs)
  File "/Users/u1166368/miniforge3/envs/esmvaltool/lib/python3.14/site-packages/ncdata/iris.py", line 52, in to_iris
    dslikes = Nc4DatasetLike(ncdata)
  File "/Users/u1166368/miniforge3/envs/esmvaltool/lib/python3.14/site-packages/ncdata/dataset_like.py", line 112, in __init__
    name: Nc4VariableLike._from_ncvariable(ncvar)
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "/Users/u1166368/miniforge3/envs/esmvaltool/lib/python3.14/site-packages/ncdata/dataset_like.py", line 258, in _from_ncvariable
    self = cls(
        ncvar=ncvar,
        datatype=dtype,
    )
  File "/Users/u1166368/miniforge3/envs/esmvaltool/lib/python3.14/site-packages/ncdata/dataset_like.py", line 251, in __init__
    array = self._maskandscale_inner_to_apparent(array)
  File "/Users/u1166368/miniforge3/envs/esmvaltool/lib/python3.14/site-packages/ncdata/dataset_like.py", line 308, in _maskandscale_inner_to_apparent
    fv = self._get_fillvalue()
  File "/Users/u1166368/miniforge3/envs/esmvaltool/lib/python3.14/site-packages/ncdata/dataset_like.py", line 298, in _get_fillvalue
    fv = netCDF4.default_fillvals[dtype_code]
```


Are there fixes for these sorts of issues elsewhere in the codebase? From a user experience standpoint, having these catalogs variously raise conversion errors is not going to be pleasant or encouraging...

bouweandela commented on tests/unit/io/test_intake_esm.py:
```diff
@@ -0,0 +1,400 @@
+            match=r"When encoding chunked arrays of datetime values, both the units and dtype must be prescribed or both must be unprescribed. Prescribing only one or the other is not currently supported. Got a units encoding of hours since 0151-01-16 12:00:00.000000 and a dtype encoding of None.",
+        ):
+            ds.to_iris(
+                xarray_open_kwargs={"decode_cf": True, "decode_times": True},
```
```suggestion
                xarray_open_kwargs={"decode_cf": False, "decode_times": False},
```
seems to avoid the issue. It looks like something in ncdata isn't quite working when you feed it decoded time coordinates, but it is more efficient not to decode the times with xarray, as iris needs them encoded anyway. Then we run into the other issue with `ncdata` failing on the code that is looking up the default fill-value for the `member_id` coordinate, that may be partially solved by https://github.com/SciTools/ncdata/pull/223.

bouweandela replied:
With the changes suggested above and here https://github.com/ESMValGroup/ESMValCore/pull/2690/files#r3073813899, this test passes for me.

bouweandela commented on esmvalcore/io/intake_esm.py:
```diff
@@ -0,0 +1,280 @@
+    def attributes(self, value: dict[str, Any]) -> None:
+        self._attributes = value
+
+    def to_iris(self, quiet: bool = True, **kwargs) -> iris.cube.CubeList:  # noqa: ANN003
```
When using this data source from a recipe, there is no way that you can pass `**kwargs` in here. Typically, these keyword arguments will depend on the data source, so I think that the best solution would be to make it possible to specify these in the `data-intake-esm-*.yml` configuration files instead.

bouweandela commented on esmvalcore/config/configurations/data-intake-esm-dkrz-cloud.yml:
```diff
@@ -0,0 +1,21 @@
+# Documentation example: intake-esm catalog (DKRZ - cloud access)
```
The DKRZ catalogs appear to be for an older version of intake-esm, probably easiest to drop them for now and add them later

bouweandela commented on esmvalcore/config/configurations/data-intake-esm-aws.yml:
```diff
@@ -0,0 +1,21 @@
+          version: "version"
+          timerange: "time_range"
+        time_separator: "/"
+        catalog: "https://<AWS_INTAKE_ESM_CATALOG_URL>"
```
Replace by the actual URL?

bouweandela commented on esmvalcore/io/intake_esm.py:
```diff
@@ -0,0 +1,280 @@
+        dataset.attrs["source_file"] = ", ".join(str(f) for f in files)
+        # Cache the attributes.
+        self.attributes = copy.deepcopy(dataset.attrs)
+        return dataset_to_iris(dataset)
```
To avoid the issue with the extra coordinates that Pangeo data has, I would suggest solving this like this for now:
```suggestion
        if (
            "pangeo" in self.catalog.esmcat.id
            and dataset.sizes.get("member_id") == 1
            and dataset.sizes.get("dcpp_init_year") == 1
        ):
            # Temporarily remove the extra dimensions in pangeo data here, until there is a more general solution available.
            dataset = dataset.isel(member_id=0, dcpp_init_year=0).drop_vars(
                ["member_id", "dcpp_init_year"],
            )
        return dataset_to_iris(dataset)
```

bouweandela commented on tests/unit/io/test_intake_esm.py:
```diff
@@ -0,0 +1,400 @@
+    results = data_source.find_data(
+        institution_id="NOAA-GFDL",
+        table_id="Amon",
+        experiment_id="piControl",
```
Use the [ESMValTool facet names](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/reference/facets.html) with the data source
```suggestion
        institute="NOAA-GFDL",
        mip="Amon",
        exp="piControl",
```

bouweandela commented:
Great to see this progressing to a usable state.

> Are there fixes for these sorts of issues elsewhere in the codebase? From a user experience standpoint, having these catalogs variously raise conversion errors is not going to be pleasant or encouraging...

We're working on a general solution for making data analysis ready, but it is not ready yet. Prototypes are available at https://github.com/ESMValGroup/fixer-prototype and https://github.com/cehbrecht/woodpecker.

> ValueError: When encoding chunked arrays of datetime values

I believe this is an issue with `ncdata`, it can be avoided by not decoding the times when loading data into xarray. See https://github.com/ESMValGroup/ESMValCore/pull/2690#discussion_r3028402553 for further detail

> *** KeyError: 'O'

Another issue with `ncdata`, it can be avoided by the code suggested here: https://github.com/ESMValGroup/ESMValCore/pull/2690#discussion_r3073813899.

charles-turner-1 committed:
Bring things back up to date & address kwargs stuff

charles-turner-1 committed:
Pre-commit

charles-turner-1 committed:
Merge branch 'main' into intake-esm

charles-turner-1 committed:
Merge branch 'main' into intake-esm
